### PR TITLE
:bug: Eventshub receiver readiness ensured

### LINKED ISSUE #270: Connection refused when sending events EventsHub receiver with external sender
The EventsHub receiver seams to be lacking in readiness processing. It is possible to deploy a receiver, and while it is reported as ready, get `Connection refused` error, while sending an event with a high chance.

Below is the current code that checks for readiness of the eventshub:

https://github.com/knative-sandbox/reconciler-test/blob/816f2192fec9b6363c2b9ed79875aa54d472eede/pkg/eventshub/resources.go#L81

Some hints:

- Maybe adding a proper readiness probe to the eventshub pod could help this situation
- It may be that the service isn't ready yet, despite the pod being ready. See: https://stackoverflow.com/a/59713538/844449

/kind bug

### PR DESCRIPTION
**Pull request by cardil**
# Changes

- :bug: Eventshub receiver readiness ensured

/kind bug

Fixes #270
Closes #271

### CODE REVIEW

File: pkg/k8s/job.go [hunk 1]
```diff
@@ -28,27 +29,39 @@ import (
...
+	jobs := kube.BatchV1().Jobs(namespace)
+	log := logging.FromContext(ctx)
 
 	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
-		job, err := client.BatchV1().Jobs(namespace).Get(context.Background(), name, metav1.GetOptions{})
+		job, err := jobs.Get(ctx, name, metav1.GetOptions{})
 		if err != nil {
 			if apierrors.IsNotFound(err) {
-				t.Log(namespace, name, "not found", err)
+				log.Debugf("%s/%s job %+v", namespace, name, err)
```
Comment (pierDipi) [inline pkg/k8s/job.go:50]: Why this? We want to have test-scoped logs.

/hold

**Review by pierDipi — CHANGES_REQUESTED** [review]
I'm facing this issue, thanks for the fix!

File: pkg/logging/context.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	return logging.WithLogger(ctx, logger())
+}
+
+func logger() *zap.SugaredLogger {
+	if log, err := zap.NewDevelopment(); err != nil {
+		panic(err)
+	} else {
+		return log.Named("test").Sugar()
+	}
+}
```
Comment (pierDipi) [inline pkg/logging/context.go:41]: The logger created here won't be bound to a `testing.T` variable, so we don't get test-scoped logs.

Comment (cardil) [reply pkg/logging/context.go:41]: That's right. The logging is all over the place (`t.Log`, `zap`, `fmt.Print`, `log`, `k8s.io/klog`) I plan to refactor it all to use [`zaptest.NewLogger`](https://pkg.go.dev/go.uber.org/zap/zaptest#NewLogger) - that will bound all messages to `testing.T`.

But, maybe we could do that in separate PR, WDYT?

Comment (pierDipi) [reply pkg/logging/context.go:41]: +1, that would be great!

We should inject the logger in the context when `MagicEnvironment.Test(...)` is called.

My only concern of this PR is that, [as per my previous comment](https://github.com/knative-sandbox/reconciler-test/pull/284#discussion_r793311633), bring a regression in terms of logging scope, so I'd say let's revert that and work on the refactoring in a separate PR.

Comment (cardil) [reply pkg/logging/context.go]: I've refactored the logging in 5768204.

Maybe we can do it in one PR?!?

Comment (cardil) [reply pkg/logging/context.go]: Hmm, on second thought, I think I'll divide this PR, and create a second one targeted solely on logging.

Comment (cardil) [reply pkg/logging/context.go:41]: Opened the #294 for logging overhoul

File: pkg/k8s/job.go (hunk 1 above)
Comment (cardil) [reply pkg/k8s/job.go]: Restored in fea507b

Moved to #294

/unhold

File: pkg/k8s/wait.go [hunk 3]
```diff
@@ -187,53 +207,162 @@ func isNotReady(t feature.T, name string) ConditionFunc {
...
+}
+
+// WaitForServiceReadyOrFail will call WaitForServiceReady and fail if error is returned.
+func WaitForServiceReadyOrFail(ctx context.Context, t feature.T, name string, readinessPath string) {
+	if err := WaitForServiceReady(ctx, t, name, readinessPath); err != nil {
+		t.Fatalf("Failed while %+v", errors.WithStack(err))
+	}
+}
+
+const ubi8Image = "registry.access.redhat.com/ubi8/ubi"
```
Comment (pierDipi) [inline pkg/k8s/wait.go:297]: Maybe we can use centos?

**Review by pierDipi — APPROVED** [review]
/lgtm

/hold in case you want to address my comments regarding the image

File: test/e2e/eventshub/receiver_readiness.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+	"knative.dev/pkg/kmeta"
+	"knative.dev/reconciler-test/pkg/environment"
+	"knative.dev/reconciler-test/pkg/eventshub"
+	"knative.dev/reconciler-test/pkg/eventshub/assert"
+	"knative.dev/reconciler-test/pkg/feature"
+	"knative.dev/reconciler-test/pkg/k8s"
+	"sigs.k8s.io/yaml"
+)
+
+const ubi8Image = "registry.access.redhat.com/ubi8/ubi"
```
[the remark this answers is not in the document]
Comment (cardil) [reply test/e2e/eventshub/receiver_readiness.go:48]: Ubi image is exactly the same legal situation as using CentOS stream or Fedora. All of the are free for use by any party (community or software vendor). I don't see a reason to switch.

From https://catalog.redhat.com/software/containers/ubi8/ubi/5c359854d70cc534b3a3784e?container-tabs=gti:

> The Red Hat Universal Base Image is free to deploy on Red Hat or non-Red Hat platforms and freely redistributable. Software vendors and community projects which build on UBI may have additional EULAs which apply to their layered software. Please refer to the End User License Agreement for the Red Hat Universal Base Image for information about use of the Red Hat Universal Base Image and associated software and source code.

[the remark this answers is not in the document]
Comment (cardil) [reply test/e2e/eventshub/receiver_readiness.go:48]: And using UBI we should avoid potential problems with docker.io download quotas.

[the remark this answers is not in the document]
Comment (pierDipi) [reply test/e2e/eventshub/receiver_readiness.go:48]: Ok, cool!

/unhold

**Comment by cardil** [comment]
:sob: Permission not propagated yet https://github.com/knative/community/pull/910

**Pull request merged**